core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1659,40 +1660,58 @@ sealed trait ZIO[-R, +E, +A]
   final def retry[R1 <: R, S](
     policy: => Schedule[R1, E, S]
   )(implicit ev: CanFail[E], trace: Trace): ZIO[R1, E, A] =
-    ZIO.suspendSucceed {
-
-      def loop(driver: Schedule.Driver[Any, R1, E, S]): ZIO[R1, E, A] =
-        self.catchAllCause { cause =>
-          cause.failureOrCause.fold(
-            e =>
-              driver
-                .next(e)
-                .foldZIO(
-                  _ => driver.last.orDie.flatMap(_ => Exit.failCause(cause)),
-                  _ => loop(driver)
-                ),
-            cause => Exit.failCause(cause)
+    policy.driver.flatMap { driver =>
+      var s: Option[A] = None
```

Comment:
Maybe? Then if the value of the ZIO was `null`, then the `whileLoop` would never terminate, but maybe it is ok to expect the value not to be null?

Reply:
While I think it's OK to make that assumption, perhaps let's keep `Option` at the moment since the penalty is relatively small (just 1 object allocation)